core/src/main/java/org/infinispan/protostream/impl/TagReaderImpl.java

```diff
@@ -1066,186 +995,75 @@ InputStream getInputStream() {
-         if (!tryFillBuffer(requestedBytes)) {
-            throw log.messageTruncated();
-         }
-      }
-
-      /**
-       * Tries to fill the buffer with at least the requested number of bytes, or more, but no more than the buffer
-       * capacity and indicates if the operation succeeded or failed either due to lack of available data in stream or
-       * by reaching the limit set with pushLimit.
-       */
-      private boolean tryFillBuffer(int requestedBytes) throws IOException {
-         if (requestedBytes + pos <= end) {
-            // all requested bytes already available; nothing to do
+         if (pos == limit) {
             return true;
          }
-
-         if (requestedBytes + bytesBeforeStart + pos > limit) {
-            // oops, should not exceed the limit
-            return false;
-         }
-
-         // slide existing data if some bytes are already consumed
-         if (pos > 0) {
-            if (end > pos) {
-               System.arraycopy(buf, pos, buf, 0, end - pos);
+         if (in instanceof PushbackInputStream) {
+            int intVal = in.read();
+            if (intVal < 0) {
+               return true;
             }
-            bytesBeforeStart += pos;
-            end -= pos;
-            pos = 0;
-         }
-
-         // fill the space at the end with data from stream
-         int read = in.read(buf, end, buf.length - end);
-         if (read <= 0) {
-            // EOF or maybe our buffer is full and we attempted to read 0 bytes
+            ((PushbackInputStream) in).unread(intVal);
             return false;
          }
-         end += read;
-
-         if (requestedBytes + bytesBeforeStart - globalLimit > 0) {
-            throw log.globalLimitExceeded();
+         in.mark(1);
+         if (in.read() < 0) {
+            return true;
          }
-
-         adjustEnd();
-         return end >= requestedBytes || end == buf.length || tryFillBuffer(requestedBytes);
+         in.reset();
+         return false;
       }
 
+
       @Override
       byte readRawByte() throws IOException {
-         if (pos == end) {
-            fillBuffer(1);
+         if (pos == limit) {
+            throw log.messageTruncated();
+         }
+         int byteValue = in.read();
+         if (byteValue < 0) {
+            throw log.messageTruncated();
          }
-         return buf[pos++];
+         pos++;
+         return (byte) byteValue;
       }
 
       @Override
       byte[] readRawByteArray(int length) throws IOException {
-         if (length > 0 && length <= end - pos) {
-            int from = pos;
+         if (length > 0 && length <= limit - pos) {
             pos += length;
-            return Arrays.copyOfRange(buf, from, pos);
+            if (pos > globalLimit) {
+               throw log.globalLimitExceeded();
+            }
+            byte[] array = new byte[length];
+            int readAmount = in.read(array);
```

Comment:
shouldn't be `readNBytes()`?
```suggestion
            int readAmount = in.readNBytes(array, 0, length);
```

Reply:
No, this way is fine. The length is implied with the provided byte[]. Unless you are thinking something else is wrong?

Reply:
Are you thinking because it may return less bytes than it should in a blocking operation?

Reply:
I can update both to use that instead to be sure.

Reply:
In Infinispan we have all the bytes available, so it shouldn't matter. But using the `readNBytes` seems more correct.